src/uaclient/uasession.cpp

```diff
@@ -328,5 +328,69 @@ UaStatus UaSession::call(
 
 }
 
+UaStatus UaSession::browse(
+        ServiceSettings&         serviceSettings,
+        const UaNodeId&          nodeToBrowse,
+        const BrowseContext&     browseContext,
+        UaByteString&            continuationPoint,
+        UaReferenceDescriptions& referenceDescriptions)
+{
+	referenceDescriptions.clear();
+
+    UA_BrowseRequest browseRequest;
+    UA_BrowseRequest_init(&browseRequest);
+
+    browseRequest.nodesToBrowseSize = 1;
+    browseRequest.nodesToBrowse = UA_BrowseDescription_new();
+    nodeToBrowse.copyTo(&browseRequest.nodesToBrowse[0].nodeId);
+    browseRequest.nodesToBrowse[0].browseDirection = UA_BROWSEDIRECTION_FORWARD; // TODO!! Important!! This setting should come from BrowseContext! It might be an oversimplificatrion.
+    browseRequest.nodesToBrowse[0].resultMask = UA_BROWSERESULTMASK_ALL;
+
+    UA_BrowseResponse browseResponse;
+    UA_BrowseResponse_init(&browseResponse);
+
+    browseResponse = UA_Client_Service_browse(m_client, browseRequest);
+
+    UaStatus finalStatusCode = browseResponse.responseHeader.serviceResult;
+
+    if (UaStatus(browseResponse.responseHeader.serviceResult).isGood())
+    {
+    	LOG(Log::INF) << "After browse (was OK): resultsSize= " << browseResponse.resultsSize;
+    	if (browseResponse.resultsSize == 1) // we asked to browse one node, so should get one result.
+    	{
+    		UA_BrowseResult& br = browseResponse.results[0];
+    		LOG(Log::DBG) << "Browse result 0, references size: " << br.referencesSize;
+    		referenceDescriptions.create(br.referencesSize);
+    		for (int j=0; j < br.referencesSize; j++)
+    		{
+    			UA_ReferenceDescription& rd = br.references[j];
+    			LOG(Log::DBG) << "... ref #" << j << ": " <<
+    					"isForward=" << rd.isForward << ", " <<
+						"refTypeId=" << UaNodeId(rd.referenceTypeId).toString().toUtf8() << ", " <<
+						"browseName=" << UaString(rd.browseName.name).toUtf8();
+    			// populate the UA-SDK counter-part now
+    			referenceDescriptions[j].IsForward = rd.isForward;
+    			referenceDescriptions[j].ReferenceTypeId = rd.referenceTypeId;
+    			referenceDescriptions[j].NodeId = UaExpandedNodeId(
+    					rd.nodeId.nodeId,
+						rd.nodeId.namespaceUri,
+						rd.nodeId.serverIndex);
+    			referenceDescriptions[j].BrowseName = rd.browseName.name;
+    		}
+    	}
+    	else
+    	{
+    		LOG(Log::ERR) << "Unexpected result size: " << browseResponse.resultsSize << ", expected was 1";
+    		finalStatusCode = UA_STATUSCODE_BADUNKNOWNRESPONSE;
+    	}
+
+    }
+
+    UA_BrowseRequest_clear(&browseRequest);
```

Comment:
These UA_blahtype_clear() functions always required (avoid leaking memory or something)?

Maybe consider wrap up the browseRequest & browseResponse in a std::unique_ptr with a custom deleter (that calls UA_blahtype_clear()) so that even if say this method explodes from the middle somewhere (e.g. exception) then those tidy up calls will be made when the stack unwinds.

Reply:
yep, these are required - in the end, open6 is C-style memory management (perhaps more elaborate due to distinct mem mgmt functions per type, but still).

Your remark is valid IMO, but is far more general. We'd need some RAII wrapper for this, I agree -- we already have one for all arrays that are used in open62541-compat (ManagedUaArray template). I can create a ticket to approach this kind of en-masse. Thanks for the suggestion.